oak_restricted_kernel/src/lib.rs

```diff
@@ -332,11 +332,44 @@ pub fn start_kernel(info: &BootParams) -> ! {
         sev_status,
     );
 
-    // We need to load the application binary before we hand the channel over to the syscalls, which
-    // expose it to the user space.
-    info!("Loading application binary...");
-    let application = payload::Application::load_raw(&mut *channel)
-        .expect("failed to load application binary from channel");
+    let application: payload::Application = {
+        let ramdisk: Option<&[u8]> = match info.hdr.ramdisk_size {
+            0 => None,
+            _ => {
+                let start_addr = {
+                    let phys_start_addr = PhysAddr::new_truncate(info.hdr.ramdisk_image.into());
+                    let pt = PAGE_TABLES.get().expect("failed to get page tables");
+                    pt.translate_physical(phys_start_addr)
+                        .expect("failed to translate physical dice address")
+                };
+
+                // Safety: the firmware ensures that the location of the ramdisk is marked as
```

Comment:
IIRC it marks the location as RAM, not reserved. I think there is some work that must happen earlier, before setting up the heap allocator to ensure that the RAM disk range is not used by the allocator. 

The wrapper must probably also be updated to make sure that there is no overlap between this range and the range that is used to lay the kernel ELF payload in memory.

Then the safety statement should mention the following:

- we rely on the firmware to ensure this range is valid and backed by physical memory
- we excluded this range from the frame allocator so it cannot be used by the heap allocator
- the wrapper checked that this range does not conflict with the memory used by the kernel for the stack BSS, data, read-only data or code sections

Reply:
thanks for context, that's helpful. In my testing it was marked as reserved, will investigate further and update!

Reply:
given how stage0 zero allocates the initrd, definitely it does not use a RAM location, confirming my earlier investigation.

https://github.com/project-oak/oak/blob/04c5ec22b2990b63c954daeea064280aa5f15b63/stage0/src/initramfs.rs#L43

Now we may not want to rely on that, other firmwares may behave differently. 

Also stage0's allocation mechanism for the initrd seems odd to me (using find_suitable_dma_address to find a range specifically, seems counterintuitive and not robust). @andrisaar am I misreading something here?

Reply:
AFAIK find_suitable_dma_address tries to find a suitable slice in the highest possible memory range marked as RAM and I don't think it then marks this range as reserved after finding it.

The idea behind this is to find a contiguous location in RAM (using the E820 table to know what memory is available) that is not used and big enough for the initial RAM disk. I think that is safer than relying on the VMMs suggested location. Do you have a more robust idea in mind?

Reply:
yes, I think it needs to be smaller than that even (perhaps 700MiB) so that the kernel also fits. But, I think that is OK for now since we don't expect initial RAM disks that are more than about 10MiB at  the moment (IIRC).

Reply:
> the wrapper checked that this range does not conflict with the memory used by the kernel for the stack BSS, data, read-only data or code sections

@conradgrobler could you elaborate on what this means? 

I'm not what the wrapper is (the application struct?), and what these refer to "stack BSS, data, read-only data or code sections" (is this the memory where the parsed elf file will be copied into?). 

Naively it seems these should not be novel concerns, afaik the ramdisk_slice is cloned into onto the heap before being passed the `Application` struct, just like an app received over the channel. (We might want to optimize this).

Reply:
> I'm not what the wrapper is (the application struct?), and what these refer to "stack BSS, data, read-only data or code sections" (is this the memory where the parsed elf file will be copied into?).

The wrapper is the `oak_restricted_kernel_wrapper`. It contains the kernel ELF file. It parses the kernel ELF file and lays the different segments (code, data etc.) into memory locations defined by the ELF file. If any of these overwrites the init ram disk, the image will be corrupted so we should fail. This happens before the kernel starts booting, so cannot be handled in the kernel itself.

Reply:
done, could you have another look? Thank you.